crates/wasmtime/src/runtime/component/func/options.rs

```diff
@@ -86,12 +86,18 @@ impl Options {
         self.store_id.assert_belongs_to(store.0.id());
 
         let realloc = self.realloc.unwrap();
+        let realloc_ty = FuncType::new(
+            store.engine(),
+            [ValType::I32, ValType::I32, ValType::I32, ValType::I32],
+            [ValType::I32],
+        );
```

Comment:
Could always just throw it somewhere in the store I guess... Is that the best approach?

Reply:
At a high level what I'd expect is:

* At `Component` compilation time the `ModuleTypesBuilder` gets the realloc signature added, no matter whether it was already there or not. This is saved in the component compilation artifacts.
* At `Component` creation time this local interned index is turned into a engine-interned index and saved in a `Component`
* Upon instantiation this engine-interned index makes its way into `Instance` which I think is referenced from functions, either that or the index is copied into functions/options/etc
* The `call_raw` function is updated to take an index instead of an object, ... somehow ... (unsure about this) ... and then avoids reifying the type each time. Either that or I suppose the `FuncType` could be stored somewhere in `Component` and make its way into `Instance`/`FuncData`.

Or something like that.

Reply:
Hm so not all `realloc` callers have a full `wasmtime::component::Instance`, a bunch just have a pointer to a `wasmtime_runtime::ComponentInstance`, and the latter can't (directly) hold a `wasmtime::FuncType` because of the artificial crate boundary between `wasmtime` and `wasmtime_runtime`. We could stuff it in a `Box<Any>` inside `wasmtime_runtime::ComponentInstance` but that is pretty unsatisfactory.

Alternatively, if we remove the dynamic type checks style of `Func::typed` that I added in this PR, then we don't need a `FuncType` for `TypedFunc::call_raw` anymore, which means we don't need realloc's `FuncType` here anymore, and `TypedFunc` and `Options` can even remain `Copy`. That does, however, mean we need an alternative way of doing typed function calls for functions that use concrete reference types, and the only alternative that I can think of at the moment (other than just not suppporting them, which is super unsatisfactory) is making `TypedFunc` implement `WasmTy`, as discussed below.